src/utils/defaultMap.js

```diff
@@ -0,0 +1,33 @@
+// @flow strict-local
+
+/**
+ * A wrapper around Map, akin to Python's `defaultdict`.
+ */
+// This could be a subclass of Map instead.  That would also mean it gets
+// all the handy methods of Map.
+//
+// The slightly tricky bit is that that would mean it could enter our Redux
+// state where the type says simply `Map`.  So `replaceRevive` would need to
+// handle it, and some picky tests might need adjustment too.
+//
+// We haven't done that yet mainly because we haven't yet had much of a use
+// case for using this outside of small local contexts where we don't need
+// the other methods of Map.
+export default class DefaultMap<K, V> {
+  map: Map<K, V> = new Map();
+  default: () => V;
+
+  constructor(default_: () => V) {
+    this.default = default_;
+  }
+
+  /** Return the value at the given key, creating it if necessary. */
+  getOrCreate(key: K): V {
+    let value = this.map.get(key);
+    if (value === undefined) {
+      value = this.default();
+      this.map.set(key, value);
```

Comment:
Huh—😮—looking at the [`defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict) doc, I guess this is what it does too.

I guess callers should be careful to make sure `default_` stays pretty steady/predictable in what it returns across the lifetime of the `DefaultMap`. Otherwise I could imagine some annoying bugs where a value at a key accidentally depends on some state at the time the key was first read, if the key was missing then.

I like how you've named this method `getOrCreate` and mentioned the "create" part in the jsdoc; I don't think this needs anything else to be clear enough about the behavior.

Looking at the examples in the doc, it seems normal to pass something like the `list` constructor. Which is very steady/predictable in what it returns. 😛

Reply:
It looks like `Immutable.Map`'s [`.get()`](https://immutable-js.com/docs/v4.0.0/Map/#get()) takes an optional `notSetValue`. I'm not sure if that's a great alternative—callers may accidentally vary what they pass for `notSetValue` with each `.get()` call and run into bugs that way. But at least there's not this odd possibility of depending on the state at first-read-if-not-set time.

Reply:
Yeah, I think the right way to use `defaultdict` always involves giving it a factory that makes something like an empty container, or returns a constant -- something that manifestly doesn't depend on any data that can change.

It'd be possible to do something twisty like a factory that ends up depending on some state, but that wouldn't be a good use case for `defaultdict` -- and I don't think it's a tempting one in practice, as I've never seen any code like that.


> It looks like `Immutable.Map`'s [`.get()`](https://immutable-js.com/docs/v4.0.0/Map/#get()) takes an optional `notSetValue`.

That does something quite different -- it doesn't alter the underlying collection (nor return a modified collection). Its Python analog would be `dict.get`, on the plain `dict` type:
https://docs.python.org/3/library/stdtypes.html#dict.get

What is more like `defaultdict` or this `DefaultMap` is the `notSetValue` parameter to `Immutable.Map#update`:
https://immutable-js.com/docs/v4.0.0/Map/#update()
Hmm, oddly that doesn't seem to have a description. But I believe the behavior is basically as if it were temporarily a DefaultMap with a factory returning `notSetValue`.

I think that `notSetValue` API choice is a reasonable compromise for the sake of having just the one `Map` type and having that feature still exist. But I think it is quite prone to bugs of this kind:

> I'm not sure if that's a much better API—callers may accidentally vary what they pass for `notSetValue` with each `.get()` call and run into bugs that way.